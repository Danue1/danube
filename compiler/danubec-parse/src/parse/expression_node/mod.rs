mod literal_expression_node;

use crate::State;

impl crate::Context {
    pub fn expression_node(&mut self) -> State {
        one_of!(self.expression_literal_node())
    }
}
