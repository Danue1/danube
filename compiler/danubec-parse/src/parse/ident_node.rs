use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn ident_node(&mut self) -> State {
        guard!(self, IDENT_KEYWORD, IDENT_NODE);
        self.finish_node();
        State::Stop
    }
}
