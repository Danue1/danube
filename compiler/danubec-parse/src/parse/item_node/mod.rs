mod enum_item_node;
mod struct_item_node;

impl crate::Context {
    pub fn item_node(&mut self) -> crate::State {
        one_of!(self.struct_item_node(), self.enum_item_node())
    }
}
