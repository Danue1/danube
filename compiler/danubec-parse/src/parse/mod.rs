mod ident_node;
mod item_node;
mod path_node;
mod type_node;

use crate::context::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::context::Context {
    #[inline]
    pub fn ast(&mut self) {
        self.start_node(SyntaxKind::AST_NODE);
        self.skip_whitespace();
        while !self.is_empty() {
            if self.item_node() == State::Continue {
                self.unexpected_token();
            }
            self.skip_whitespace();
        }
        self.finish_node();
    }

    fn unexpected_token(&mut self) {
        self.start_node(SyntaxKind::ERROR);
        self.bump();
        self.finish_node();
    }
}
