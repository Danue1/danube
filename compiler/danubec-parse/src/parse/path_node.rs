use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn path_node(&mut self) -> State {
        guard!(self, IDENT_KEYWORD, PATH_NODE, ident_node);

        while !self.is_eof() {
            if expect!(self, SyntaxKind::COLON) {
                if expect!(self, SyntaxKind::COLON) {
                    self.skip_whitespace();
                    self.ident_node();
                } else {
                    self.unexpected_token();
                }
            } else {
                break;
            }
        }

        self.finish_node();
        State::Stop
    }
}
