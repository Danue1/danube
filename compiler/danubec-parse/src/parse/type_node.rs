use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn type_node(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.unnamed_type_kind_node() == State::Continue
            && self.path_type_kind_node() == State::Continue
        {
            State::Continue
        } else {
            self.start_node_at(checkpoint, SyntaxKind::TYPE_NODE);
            self.finish_node();
            State::Stop
        }
    }

    // ()
    // (type1)
    // (type1, type2)
    pub fn unnamed_type_kind_node(&mut self) -> State {
        guard!(self, LEFT_PAREN, UNNAMED_TYPE_KIND_NODE);
        self.skip_whitespace();

        while !self.is_eof() {
            if expect!(self, SyntaxKind::RIGHT_PAREN) {
                break;
            }

            let checkpoint = self.checkpoint();
            if self.type_node() == State::Continue {
                self.unexpected_token();
            } else {
                self.start_node_at(checkpoint, SyntaxKind::UNNAMED_TYPE_KIND_NODE);
                self.finish_node();
            }
            self.skip_whitespace();
            if expect!(self, SyntaxKind::COMMA) {
                self.skip_whitespace();
            }
        }

        self.finish_node();
        State::Stop
    }

    // Foo
    // Foo::Bar
    pub fn path_type_kind_node(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.path_node() == State::Continue {
            return State::Continue;
        }

        self.start_node_at(checkpoint, SyntaxKind::PATH_TYPE_KIND_NODE);
        self.finish_node();
        State::Stop
    }
}
