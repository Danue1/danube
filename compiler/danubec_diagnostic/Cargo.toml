[package]
name = "danubec_diagnostic"
version = "0.1.0"
edition = "2024"

[dependencies]
