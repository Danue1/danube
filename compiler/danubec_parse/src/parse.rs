use crate::{
    event::Event,
    grammar::{Context, krate},
};
use danubec_diagnostic::Diagnostic;
use danubec_lex::lex;
use danubec_syntax::{GreenNodeBuilder, SyntaxKind, SyntaxNode};

pub fn parse(source: &str, diagnostic: &mut Diagnostic) -> SyntaxNode {
    let mut parse = |tokens: &[_]| {
        let tokens: Vec<_> = tokens.iter().map(|&(kind, _)| kind).collect();
        let mut context = Context::new(&tokens, diagnostic);
        krate(&mut context);

        context.finish()
    };

    let tokens = lex(source);
    let events = parse(&tokens);

    build(&tokens, events)
}

fn build(mut tokens: &[(SyntaxKind, &str)], mut events: Vec<Event>) -> SyntaxNode {
    let mut builder = GreenNodeBuilder::new();

    macro_rules! advance {
        () => {
            if let Some(&(kind, text)) = tokens.get(0) {
                builder.token(kind.into(), text);
                tokens = &tokens[1..];
            }
        };
    }

    macro_rules! trivia {
        () => {
            loop {
                match tokens.get(0) {
                    Some((kind, _)) if kind.at_trivia() => advance!(),
                    _ => break,
                }
            }
        };
    }

    for index in 0..events.len() {
        match std::mem::replace(&mut events[index], Event::Placeholder) {
            event @ Event::Start { forward_parent, .. }
            | event @ Event::Expire { forward_parent } => {
                let mut forward_parents = match event {
                    Event::Start { kind, .. } => vec![kind],
                    _ => vec![],
                };
                let mut index = index;
                let mut forward_parent = forward_parent;

                while let Some(fp) = forward_parent {
                    index += fp;
                    match std::mem::replace(&mut events[index], Event::Placeholder) {
                        Event::Start {
                            kind,
                            forward_parent: fp,
                        } => {
                            forward_parents.push(kind);
                            forward_parent = fp;
                        }
                        Event::Expire { forward_parent: fp } => {
                            forward_parent = fp;
                        }
                        _ => {
                            //
                        }
                    }
                }

                for kind in forward_parents.into_iter().rev() {
                    builder.start_node(kind.into());
                }

                trivia!();
            }
            Event::Token => {
                advance!();
                trivia!();
            }
            Event::End => builder.finish_node(),
            Event::Placeholder => {
                //
            }
        }
    }

    SyntaxNode::new_root(builder.finish())
}
