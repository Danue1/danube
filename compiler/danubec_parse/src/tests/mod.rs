use crate::parse;
use danubec_diagnostic::Diagnostic;

#[test]
fn top_level_attribute() {
    let source = r#"
#![a("value")]
#![a(key)]
#![a(key1, key2)]
#![a(key = "value")]
#![a::b(key = "value")]"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn attribute() {
    let source = r#"
#[a("value")]
struct Foo;"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn zero_sized_struct() {
    let source = r#"
struct Foo;"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn named_struct() {
    let source = r#"
struct Foo { }
struct Bar {
    a: i32,
    b: String,
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn unnamed_struct() {
    let source = r#"
struct Foo();
struct Bar(());
struct Baz(i32, String);"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn mixed_enum() {
    let source = r#"
enum Foo {}
enum Bar {
    A,
    B { a: i32, b: String },
    C(i32, String),
    D = 42,
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn module() {
    let source = r#"
mod foo;
mod bar {
    struct Foo;
    enum Bar {
        A,
        B { a: i32, b: String },
        C(i32, String),
        D = 42,
    }
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn constant() {
    let source = r#"
const FOO: i32 = 42;
const BAR: str = "Hello, world!";"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn static_variable() {
    let source = r#"
static FOO: i32 = 42;
static BAR: str = "Hello, world!";"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn type_alias() {
    let source = r#"
type Foo = i32;
type Bar = (i32, String);"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn function() {
    let source = r#"
fn foo();
fn bar() {}
fn baz(a: i32, b: String) -> (i32, String) {
    (a, b)
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    dbg!(&diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn trait_definition() {
    let source = r#"
trait Foo { }
trait Bar {
    fn foo();
    const BAR: i32;
    type Baz;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn anonymous_implement() {
    let source = r#"
impl Foo { }
impl Foo {
    fn foo() {}
    const BAR: i32 = 42;
    type Baz = i32;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn targeted_implement() {
    let source = r#"
impl Foo for Bar { }
impl Foo for Bar {
    fn foo() {}
    const BAR: i32 = 42;
    type Baz = i32;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn use_definition() {
    let source = r#"
use {};
use {{},*,foo,bar as baz,::{},::*,::foo,::bar as baz,};
use *;
use foo;
use foo as bar;
use foo::{};
use foo::{*,foo,bar as baz,{},::*,::foo,::bar as baz,::{}};

use ::{};
use ::{{},*,foo,bar as baz,::{},::*,::foo,::bar as baz,};
use ::*;
use ::foo;
use ::foo as bar;
use ::foo::{};
use ::foo::{*,foo,bar as baz,{},::*,::foo,::bar as baz,::{}};
"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    dbg!(&diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}
