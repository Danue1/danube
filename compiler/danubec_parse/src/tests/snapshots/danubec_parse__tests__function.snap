---
source: compiler/danubec_parse/src/tests/mod.rs
expression: node
---
Some(
    Krate {
        attributes: [],
        definitions: [
            Definition {
                attributes: [],
                visibility: Private,
                kind: Function {
                    name: Identifier {
                        name: Ident {
                            hash: Hash64(
                                4506850079084802999,
                            ),
                            raw: "foo",
                        },
                        span: SyntaxNodePtr {
                            kind: IDENTIFIER_NODE,
                            range: 4..7,
                        },
                    },
                    type_parameters: [],
                    parameters: [],
                    return_type: None,
                    type_bounds: [],
                    body: None,
                },
                span: SyntaxNodePtr {
                    kind: DEFINITION_NODE,
                    range: 1..11,
                },
            },
            Definition {
                attributes: [],
                visibility: Private,
                kind: Function {
                    name: Identifier {
                        name: Ident {
                            hash: Hash64(
                                3676438629107045207,
                            ),
                            raw: "bar",
                        },
                        span: SyntaxNodePtr {
                            kind: IDENTIFIER_NODE,
                            range: 14..17,
                        },
                    },
                    type_parameters: [],
                    parameters: [],
                    return_type: None,
                    type_bounds: [],
                    body: Some(
                        [],
                    ),
                },
                span: SyntaxNodePtr {
                    kind: DEFINITION_NODE,
                    range: 11..23,
                },
            },
            Definition {
                attributes: [],
                visibility: Private,
                kind: Function {
                    name: Identifier {
                        name: Ident {
                            hash: Hash64(
                                16260972211344176173,
                            ),
                            raw: "baz",
                        },
                        span: SyntaxNodePtr {
                            kind: IDENTIFIER_NODE,
                            range: 26..29,
                        },
                    },
                    type_parameters: [],
                    parameters: [
                        FunctionParameter {
                            attributes: [],
                            pattern: Pattern {
                                kind: Path {
                                    path: Path {
                                        segments: [
                                            PathSegment {
                                                kind: Identifier(
                                                    Identifier {
                                                        name: Ident {
                                                            hash: Hash64(
                                                                8186225505942432243,
                                                            ),
                                                            raw: "a",
                                                        },
                                                        span: SyntaxNodePtr {
                                                            kind: IDENTIFIER_NODE,
                                                            range: 30..31,
                                                        },
                                                    },
                                                ),
                                                span: SyntaxNodePtr {
                                                    kind: PATH_SEGMENT_IDENTIFIER_NODE,
                                                    range: 30..31,
                                                },
                                            },
                                        ],
                                        span: SyntaxNodePtr {
                                            kind: PATH_NODE,
                                            range: 30..31,
                                        },
                                    },
                                },
                                span: SyntaxNodePtr {
                                    kind: PATH_PATTERN_NODE,
                                    range: 30..31,
                                },
                            },
                            type: TypeExpression {
                                kind: Path {
                                    path: Path {
                                        segments: [
                                            PathSegment {
                                                kind: Identifier(
                                                    Identifier {
                                                        name: Ident {
                                                            hash: Hash64(
                                                                1851586956637574266,
                                                            ),
                                                            raw: "i32",
                                                        },
                                                        span: SyntaxNodePtr {
                                                            kind: IDENTIFIER_NODE,
                                                            range: 33..36,
                                                        },
                                                    },
                                                ),
                                                span: SyntaxNodePtr {
                                                    kind: PATH_SEGMENT_IDENTIFIER_NODE,
                                                    range: 33..36,
                                                },
                                            },
                                        ],
                                        span: SyntaxNodePtr {
                                            kind: PATH_NODE,
                                            range: 33..36,
                                        },
                                    },
                                },
                                span: SyntaxNodePtr {
                                    kind: PATH_TYPE_NODE,
                                    range: 33..36,
                                },
                            },
                            span: SyntaxNodePtr {
                                kind: FUNCTION_PARAMETER_NODE,
                                range: 30..36,
                            },
                        },
                        FunctionParameter {
                            attributes: [],
                            pattern: Pattern {
                                kind: Path {
                                    path: Path {
                                        segments: [
                                            PathSegment {
                                                kind: Identifier(
                                                    Identifier {
                                                        name: Ident {
                                                            hash: Hash64(
                                                                16993177596579750922,
                                                            ),
                                                            raw: "b",
                                                        },
                                                        span: SyntaxNodePtr {
                                                            kind: IDENTIFIER_NODE,
                                                            range: 38..39,
                                                        },
                                                    },
                                                ),
                                                span: SyntaxNodePtr {
                                                    kind: PATH_SEGMENT_IDENTIFIER_NODE,
                                                    range: 38..39,
                                                },
                                            },
                                        ],
                                        span: SyntaxNodePtr {
                                            kind: PATH_NODE,
                                            range: 38..39,
                                        },
                                    },
                                },
                                span: SyntaxNodePtr {
                                    kind: PATH_PATTERN_NODE,
                                    range: 38..39,
                                },
                            },
                            type: TypeExpression {
                                kind: Path {
                                    path: Path {
                                        segments: [
                                            PathSegment {
                                                kind: Identifier(
                                                    Identifier {
                                                        name: Ident {
                                                            hash: Hash64(
                                                                16440811947285667152,
                                                            ),
                                                            raw: "String",
                                                        },
                                                        span: SyntaxNodePtr {
                                                            kind: IDENTIFIER_NODE,
                                                            range: 41..47,
                                                        },
                                                    },
                                                ),
                                                span: SyntaxNodePtr {
                                                    kind: PATH_SEGMENT_IDENTIFIER_NODE,
                                                    range: 41..47,
                                                },
                                            },
                                        ],
                                        span: SyntaxNodePtr {
                                            kind: PATH_NODE,
                                            range: 41..47,
                                        },
                                    },
                                },
                                span: SyntaxNodePtr {
                                    kind: PATH_TYPE_NODE,
                                    range: 41..47,
                                },
                            },
                            span: SyntaxNodePtr {
                                kind: FUNCTION_PARAMETER_NODE,
                                range: 38..47,
                            },
                        },
                    ],
                    return_type: Some(
                        TypeExpression {
                            kind: Tuple {
                                elements: [],
                            },
                            span: SyntaxNodePtr {
                                kind: TUPLE_TYPE_NODE,
                                range: 52..66,
                            },
                        },
                    ),
                    type_bounds: [],
                    body: Some(
                        [
                            Statement {
                                kind: Expression {
                                    value: Expression {
                                        kind: Tuple {
                                            elements: [
                                                Expression {
                                                    kind: Path {
                                                        path: Path {
                                                            segments: [
                                                                PathSegment {
                                                                    kind: Identifier(
                                                                        Identifier {
                                                                            name: Ident {
                                                                                hash: Hash64(
                                                                                    8186225505942432243,
                                                                                ),
                                                                                raw: "a",
                                                                            },
                                                                            span: SyntaxNodePtr {
                                                                                kind: IDENTIFIER_NODE,
                                                                                range: 73..74,
                                                                            },
                                                                        },
                                                                    ),
                                                                    span: SyntaxNodePtr {
                                                                        kind: PATH_SEGMENT_IDENTIFIER_NODE,
                                                                        range: 73..74,
                                                                    },
                                                                },
                                                            ],
                                                            span: SyntaxNodePtr {
                                                                kind: PATH_NODE,
                                                                range: 73..74,
                                                            },
                                                        },
                                                    },
                                                    span: SyntaxNodePtr {
                                                        kind: PATH_EXPRESSION_NODE,
                                                        range: 73..74,
                                                    },
                                                },
                                                Expression {
                                                    kind: Path {
                                                        path: Path {
                                                            segments: [
                                                                PathSegment {
                                                                    kind: Identifier(
                                                                        Identifier {
                                                                            name: Ident {
                                                                                hash: Hash64(
                                                                                    16993177596579750922,
                                                                                ),
                                                                                raw: "b",
                                                                            },
                                                                            span: SyntaxNodePtr {
                                                                                kind: IDENTIFIER_NODE,
                                                                                range: 76..77,
                                                                            },
                                                                        },
                                                                    ),
                                                                    span: SyntaxNodePtr {
                                                                        kind: PATH_SEGMENT_IDENTIFIER_NODE,
                                                                        range: 76..77,
                                                                    },
                                                                },
                                                            ],
                                                            span: SyntaxNodePtr {
                                                                kind: PATH_NODE,
                                                                range: 76..77,
                                                            },
                                                        },
                                                    },
                                                    span: SyntaxNodePtr {
                                                        kind: PATH_EXPRESSION_NODE,
                                                        range: 76..77,
                                                    },
                                                },
                                            ],
                                        },
                                        span: SyntaxNodePtr {
                                            kind: TUPLE_EXPRESSION_NODE,
                                            range: 72..79,
                                        },
                                    },
                                },
                                span: SyntaxNodePtr {
                                    kind: EXPRESSION_STATEMENT_NODE,
                                    range: 72..79,
                                },
                            },
                        ],
                    ),
                },
                span: SyntaxNodePtr {
                    kind: DEFINITION_NODE,
                    range: 23..80,
                },
            },
        ],
        children: {},
    },
)
